signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,256 @@
+from .caching import get_cache
+from .errors import MetadataError
+
+
+class FileBufferedCollection(BufferedCollection):
+    """Implement buffering for SyncedCollections with file-based backends.
+
+    All file-based backends can use the same set of integrity checks prior to a
+    buffer flush. This class standardizes that protocol.
+    """
+
+    _cache = get_cache()
+    _cached_collections: Dict[int, BufferedCollection] = {}
+    _BUFFER_CAPACITY = 32 * 2 ** 20  # 32 MB
+    _CURRENT_BUFFER_SIZE = 0
+
+    def __init__(self, filename, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._filename = filename
+
+    @classmethod
+    def get_buffer_capacity(cls):
+        """Get the current buffer capacity."""
+        return cls._BUFFER_CAPACITY
+
+    @classmethod
+    def set_buffer_capacity(cls, new_capacity):
+        """Update the buffer capacity."""
+        cls._BUFFER_CAPACITY = new_capacity
+
+    @classmethod
+    def get_current_buffer_size(cls):
+        """Get the total amount of data currently stored in the buffer."""
+        return FileBufferedCollection._CURRENT_BUFFER_SIZE
+
+    @staticmethod
+    def _hash(blob):
+        """Calculate and return the md5 hash value for the file data.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be hashed.
+
+        Returns
+        -------
+        str
+            The md5 hash of the input bytes.
+        """
+        if blob is not None:
+            m = hashlib.md5()
+            m.update(blob)
+            return m.hexdigest()
+
+    def _get_file_metadata(self):
+        """Return metadata of file.
+
+        Returns
+        -------
+        Tuple[int, float]
+            The size and last modification time of the associated file.
+        """
+        try:
+            metadata = os.stat(self._filename)
+            return metadata.st_size, metadata.st_mtime
+        except OSError as error:
+            if error.errno != errno.ENOENT:
+                raise
+
+    def _flush(self, force=False):
+        """Save buffered changes to the underlying file.
+
+        Parameters
+        ----------
+        force : bool
+            If True, force a flush even in buffered mode (defaults to False).
+        """
+        if not self._is_buffered or force:
+            try:
+                cached_data = self._cache[self._filename]
+            except KeyError:
+                # There are valid reasons for nothing to be in the cache (the
+                # object was never actually accessed during global buffering,
+                # multiple collections pointing to the same file, etc).
+                pass
+            else:
+                blob = json.dumps(self.to_base()).encode()
+
+                # If the contents have not been changed since the initial read,
+                # we don't need to rewrite it.
+                if self._hash(blob) != cached_data["contents"]:
```

Comment:
This is a bug, right? 🐛  The hash should not equal the cache's _contents_. We store a `"hash"` key specifically for this purpose and it's not used elsewhere.
```suggestion
                if self._hash(blob) != cached_data["hash"]:
```

Reply:
Yep, bug.